File: go/pkg/parser/struct_declaration_visitor.go
package parser

import "github.com/mojo-lang/lang/go/pkg/lang"

type StructDeclarationVisitor struct {
	*BaseMojoParserVisitor

	StructDecl *lang.StructDecl
}

func NewStructDeclarationVisitor() *StructDeclarationVisitor {
	visitor := &StructDeclarationVisitor{}
	visitor.StructDecl = &lang.StructDecl{}
	visitor.StructDecl.Type = &lang.StructType{}
	return visitor
}

func (s *StructDeclarationVisitor) VisitStructDeclaration(ctx *StructDeclarationContext) interface{} {
	s.StructDecl.Document = GetDocument(ctx.Document())
	s.StructDecl.Attributes = GetAttributes(ctx.Attributes())

	structName := ctx.StructName()
	if structName != nil {
		s.StructDecl.Name = structName.Accept(s).(string)
	} else {
		// error happened
	}

	genericParameters := ctx.GenericParameterClause()
	if genericParameters != nil {
		visitor := NewGenericParametersVisitor()
		s.StructDecl.GenericParameters = genericParameters.Accept(visitor).([]*lang.GenericParameter)
	}

	inheritances := ctx.TypeInheritanceClause()
	if inheritances != nil {
		visitor := NewTypeInheritancesVisitor()
		s.StructDecl.Type.Inherits = inheritances.Accept(visitor).([]*lang.NominalType)
	}

	body := ctx.StructBody()
	if body != nil {
		body.Accept(s)
	}

	return s.StructDecl
}

func (s *StructDeclarationVisitor) VisitStructName(ctx *StructNameContext) interface{} {
	return ctx.GetText()
}

func (s *StructDeclarationVisitor) VisitStructBody(ctx *StructBodyContext) interface{} {
	members := ctx.StructMembers()
	if members != nil {
		members.Accept(s)
	}
	return true
}

func (s *StructDeclarationVisitor) VisitStructMembers(ctx *StructMembersContext) interface{} {
	members := ctx.AllStructMember()
	for _, member := range members {
		member.Accept(s)
	}
	return true
}

func (s *StructDeclarationVisitor) VisitStructMember(ctx *StructMemberContext) interface{} {
	this := s.StructDecl
	structCtx := ctx.StructDeclaration()
	if structCtx != nil {
		visitor := NewStructDeclarationVisitor()
		structDecl := structCtx.Accept(visitor).(*lang.StructDecl)
		if structCtx != nil {
			this.StructTypeDecls = append(this.StructTypeDecls, structDecl)
		}
	}

	enumCtx := ctx.EnumDeclaration()
	if enumCtx != nil {
		visitor := NewEnumDeclarationVisitor()
		enumDecl := enumCtx.Accept(visitor).(*lang.EnumDecl)
		if enumDecl != nil {
			this.EnumTypeDecls = append(this.EnumTypeDecls, enumDecl)
		}
	}

	memberCtx := ctx.StructMemberDeclaration()
	if memberCtx != nil {
		visitor := NewValueDeclarationVisitor()
		memberDecl := memberCtx.Accept(visitor).(*lang.ValueDecl)
		if memberDecl != nil {
			this.GetType().Fields = append(this.GetType().Fields, memberDecl)
		}
	}

	return true
}

File: java/src/main/java/org/mojolang/mojo/parser/syntax/StructDeclarationVisitor.java
package org.mojolang.mojo.parser.syntax;

import com.google.protobuf.GeneratedMessageV3;
import org.mojolang.mojo.lang.EnumDecl;
import org.mojolang.mojo.lang.StructDecl;
import org.mojolang.mojo.lang.StructType;
import org.mojolang.mojo.lang.ValueDecl;

import java.util.List;

public class StructDeclarationVisitor extends MojoBaseVisitor<com.google.protobuf.GeneratedMessageV3> {
    private StructDecl.Builder builder_ = StructDecl.newBuilder();
    private StructType.Builder typeBuilder_ = StructType.newBuilder();

    @Override
    public GeneratedMessageV3 visitStruct_declaration(MojoParser.Struct_declarationContext ctx) {
        visitDocument(ctx.document());
        visitAttributes(ctx.attributes());
        visitStruct_name(ctx.struct_name());
        visitGeneric_parameter_clause(ctx.generic_parameter_clause());
        visitType_inheritance_clause(ctx.type_inheritance_clause());
        visitStruct_body(ctx.struct_body());

        builder_.setType(typeBuilder_);
        return builder_.build();
    }

    @Override
    public GeneratedMessageV3 visitDocument(MojoParser.DocumentContext ctx) {
        if (ctx != null) {
            builder_.setDocument(Util.visitDocument(ctx));
        }
        return null;
    }

    @Override
    public GeneratedMessageV3 visitAttributes(MojoParser.AttributesContext ctx) {
        if (ctx != null) {
            builder_.addAllAttributes(Util.visitAttributes(ctx));
        }
        return null;
    }

    @Override
    public GeneratedMessageV3 visitStruct_name(MojoParser.Struct_nameContext ctx) {
        builder_.setName(ctx.type_name().Type_name().getText());
        return null;
    }

    @Override
    public GeneratedMessageV3 visitGeneric_parameter_clause(MojoParser.Generic_parameter_clauseContext ctx) {
        if (ctx != null) {
            builder_.addAllGenericParameters(Util.visitGenericParameters(ctx));
        }
        return null;
    }

    @Override
    public GeneratedMessageV3 visitType_inheritance_clause(MojoParser.Type_inheritance_clauseContext ctx) {
        if (ctx != null) {
        typeBuilder_.addAllInherits(Util.visitTypeInheritances(ctx));
        }
        return null;
    }

    @Override
    public GeneratedMessageV3 visitStruct_body(MojoParser.Struct_bodyContext ctx) {
        if (ctx == null) {
            return null;
        }

        List<MojoParser.Struct_memberContext> members = ctx.struct_member();
        for (MojoParser.Struct_memberContext member : members) {
            MojoParser.Struct_declarationContext structCtx = member.struct_declaration();
            if (structCtx != null) {
                StructDeclarationVisitor visitor = new StructDeclarationVisitor();
                StructDecl struct = (StructDecl)visitor.visitStruct_declaration(structCtx);
                builder_.addStructTypeDecls(struct);
            }

            MojoParser.Enum_declarationContext enumCtx = member.enum_declaration();
            if (enumCtx != null) {
                EnumDeclarationVisitor visitor = new EnumDeclarationVisitor();
                EnumDecl declaration = (EnumDecl)visitor.visitEnum_declaration(enumCtx);
                builder_.addEnumTypeDecls(declaration);
            }

            MojoParser.Struct_member_declarationContext memberCtx = member.struct_member_declaration();
            if (memberCtx != null) {
                ValueDeclarationVisitor visitor = new ValueDeclarationVisitor();
                ValueDecl value = (ValueDecl)visitor.visitStruct_member_declaration(memberCtx);
                typeBuilder_.addFields(value);
            }
        }
        return null;
    }
}
